### PR TITLE
[WIP] AFS listener reconnection handling

### PR DESCRIPTION
**Pull request by yasmas95**

### CODE REVIEW

**Comment by yasmas95** [comment]
The client tries to connect according to the following logic :
The reconnection attempts would occur after 5", then 10", 20", 40", 1'20", 2'40" .... etc incresing until a maximum of one hour interval.
Test done manually.

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java [hunk 1]
```diff
@@ -27,6 +27,14 @@
 
     private boolean secure;
 
+    private boolean autoreconnect;
+
+    private int reconnectionInitialInterval;
+
+    private int reconnectionIntervalMutiplier;
+
+    private int reconnectionMaxInterval;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: It should be easier to configure if we have a « maxRetry » or « MaxAttemps » instead of a « MaxInterval ».

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java [hunk 2]
```diff
@@ -92,6 +111,42 @@ public RemoteServiceConfig setSecure(boolean secure) {
         return this;
     }
 
+    public boolean isAutoreconnect() {
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: refactor at least to isAutoReconnect (auto and reconnect are two different words). Maybe you should had « Enabled »

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java [hunk 3]
```diff
@@ -27,6 +27,14 @@
 
     private boolean secure;
 
+    private boolean autoreconnect;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: Refactor at least to autoReconnect (auto and reconnect are two different words). Maybe it should be autoReconnectionEnabled

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java [hunk 4]
```diff
@@ -27,6 +27,14 @@
...
+
+    private int reconnectionInitialInterval;
+
+    private int reconnectionIntervalMutiplier;
+
+    private int reconnectionMaxInterval;
+
     public RemoteServiceConfig(String hostName, String appName, int port, boolean secure) {
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: Add javadoc to all new methods to at least specify the units. Add a small description to this class to explain how do you use the new fields.

File: afs/afs-ws/afs-ws-storage/pom.xml [hunk 5]
```diff
@@ -78,6 +78,7 @@
             <artifactId>slf4j-simple</artifactId>
             <scope>test</scope>
         </dependency>
+
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/pom.xml]: Revert this change

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 6]
```diff
@@ -27,9 +36,17 @@
 
     private final WeakListenerList<AppStorageListener> listeners;
 
-    public NodeEventClient(String fileSystemName, WeakListenerList<AppStorageListener> listeners) {
+    private final RemoteServiceConfig config;
+
+    private int reconnectionInterval;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: This is redundant with the configuration?

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 7]
```diff
@@ -27,9 +36,17 @@
...
+
+    private static int counter = 0;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Should it really be a static variable? Why this counter is shared by all NodeEventClient instances?

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 8]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
         }
     }
 
+
     @OnClose
-    public void onClose(Session session) {
-        LOGGER.trace("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);
+    public void onClose(Session session, CloseReason closeReason) {
+
+        if (closeReason.getCloseCode().getCode() == 1006) {
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Is there an existing constant instead of 1006? What this error code means? Is there other error codes you should manage?

What do we do if the error code is not 1006

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 9]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+            URI wsUri = null;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Move this variable in the try block

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 10]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+            if (this.reconnectionInterval > config.getReconnectionMaxInterval()) {
+                LOGGER.warn("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);
+                return;
+            }
+            if (counter >= 1) {
+                this.reconnectionInterval = this.reconnectionInterval * config.getReconnectionIntervalMutiplier();
+            } else {
+                this.reconnectionInterval = config.getReconnectionInitialInterval();
+            }
+            WebSocketContainer container;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Same remark here

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 11]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+            try {
+
+                sleep(reconnectionInterval * 1000L);
+                wsUri = RemoteListenableAppStorage.getWebSocketUri(config.getRestUri());
+                URI endPointUri = URI.create(wsUri + "/messages/" + AfsRestApi.RESOURCE_ROOT + "/" +
+                        AfsRestApi.VERSION + "/node_events/" + fileSystemName);
+                LOGGER.debug("Connecting to node event websocket at {}", endPointUri);
+                container = ContainerProvider.getWebSocketContainer();
+                container.connectToServer(this, endPointUri);
+                resetComptor();
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Rename to resetCounter.

I don’t think counter has to be a static variable: should we have a while loop here that tries to reconnect?!

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 12]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+            } catch (DeploymentException e) {
+                if (e.getCause().getCause() instanceof ConnectException) {
+                    LOGGER.error(e.getCause().getCause().getMessage(), e);
+                    updateCounter();
+                    this.onClose(session, closeReason);
+                }
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: What do we do if the cause is different?

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 13]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+                throw new UncheckedIOException(e);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new UncheckedInterruptedException(e);
+            }
+        }
+    }
+
+    public static void updateCounter() {
+        counter = counter + 1;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: This is not thread-safe

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 14]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+    }
+
+    public static void resetComptor() {
+        counter = 0;
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: This is not thread-safe

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 15]
```diff
@@ -52,8 +69,52 @@ public void onError(Throwable t) {
...
+                throw new UncheckedInterruptedException(e);
+            }
+        }
+    }
+
+    public static void updateCounter() {
+        counter = counter + 1;
+    }
+
+    public static void resetComptor() {
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Refactor to resetCounter

Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: The counter attribute is replaced by a local variable so the method resetComptor() is deleted

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 14 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: The counter attribute is replaced by a local variable so the method resetComptor() is deleted

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 11 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: The counter attribute is replaced by a local variable in the method onClose()

File: afs/afs-ws/afs-ws-storage/pom.xml (hunk 5 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/pom.xml]: Ok, it is done

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java (hunk 3 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: Ok, it is done

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java (hunk 2 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: Ok, it is done, the new method is isAutoReconnectionEnabled

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 10 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It is done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 9 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: it is done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 7 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: The counter attribute is replaced by a local variable in method onClose()

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java (hunk 4 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: it is done 
https://github.com/po605-AFS-listener-reconnection-handlingwsybl/powsybl.github.io/pull/45/files

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java (hunk 1 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: Added intervalMax. see the link
 docs/configuration/modules/remote-service.md

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 12 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done by sending the log below
LOGGER.warn("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);

**Comment by geofjamg** [comment]
Are we  sure in the  case of normal close of storage client, we don't try to reconnect?

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 16]
```diff
@@ -54,6 +70,57 @@ public void onError(Throwable t) {
...
+            }
+        }
     }
+
+    private void updateReconnectionInterval(int counter) {
+        if (counter > 0) {
+            if (this.reconnectionInterval > config.getReconnectionMaxInterval()) {
+                this.reconnectionInterval = config.getReconnectionMaxInterval();
+            } else {
+                this.reconnectionInterval = this.reconnectionInterval * config.getReconnectionIntervalMutiplier();
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: This may exceed the ReconnectionMaxInterval. We should check against ReconnectionMaxInterval after the multiplication.

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 17]
```diff
@@ -27,9 +37,15 @@
 
     private final WeakListenerList<AppStorageListener> listeners;
 
-    public NodeEventClient(String fileSystemName, WeakListenerList<AppStorageListener> listeners) {
+    private final RemoteServiceConfig config;
+
+    private int reconnectionInterval;
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: I would rename to currentReconnectionInterval in order to remove the ambiguity.

Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: Why is this needed here? Couldn't this reconnectionInterval be just left local to the onClose method just like the counter ?

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 18]
```diff
@@ -54,6 +70,57 @@ public void onError(Throwable t) {
...
+            if (dateReconnection - dateInitial > config.getReconnectionMax()) {
+                LOGGER.warn("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);
+                return;
+            }
+            updateReconnectionInterval(counter);
+            try {
+                URI wsUri = RemoteListenableAppStorage.getWebSocketUri(config.getRestUri());
+                URI endPointUri = URI.create(wsUri + "/messages/" + AfsRestApi.RESOURCE_ROOT + "/" +
+                        AfsRestApi.VERSION + "/node_events/" + fileSystemName);
+                LOGGER.debug("Connecting to node event websocket at {}", endPointUri);
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: this should be info

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 19]
```diff
@@ -54,6 +70,57 @@ public void onError(Throwable t) {
...
+                WebSocketContainer container = ContainerProvider.getWebSocketContainer();
+                container.connectToServer(this, endPointUri);
+                toReconnection = false;
+
+            } catch (DeploymentException e) {
+                if (e.getCause().getCause() instanceof ConnectException) {
+                    try {
+                        LOGGER.error(e.getCause().getCause().getMessage(), e);
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: this should be a warning without stack trace. The warning message should mention when the next reconnection attempt will be made

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java [hunk 20]
```diff
@@ -54,6 +70,57 @@ public void onError(Throwable t) {
...
     public void onClose(Session session) {
-        LOGGER.trace("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);
+        int counter = 0;
+        boolean toReconnection = true;
+        long dateInitial = Instant.now().getEpochSecond();
+        while (toReconnection) {
+
+            long dateReconnection = Instant.now().getEpochSecond();
+            if (dateReconnection - dateInitial > config.getReconnectionMax()) {
+                LOGGER.warn("Node event websocket session '{}' closed for file system '{}'", session.getId(), fileSystemName);
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: I would move this to error level, mentionning reconnection maximum timeout reached.

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java [hunk 21]
```diff
@@ -27,6 +27,16 @@
 
     private boolean secure;
 
+    private boolean autoreconnectEnabled;
+
+    private int reconnectionInitialInterval;
+
+    private int reconnectionIntervalMutiplier;
+
+    private int reconnectionMaxInterval;
```
Comment (jeandemanged) [inline afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: I would rename to reconnectionTimeout

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 16 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 17 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 18 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 19 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

File: afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java (hunk 20 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-storage/src/main/java/com/powsybl/afs/ws/storage/NodeEventClient.java]: It's done

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java (hunk 21 above)
Comment (yasmas95) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfig.java]: It's done

**Comment by mathbagu** [comment]
AFS have been moved to powsybl-afs
